index.d.ts

```diff
@@ -1,35 +1,54 @@
-/**
-Run some code when the process exits.
+type ExitHook = {
+	/**
+	Run some code when the process exits.
 
-The `process.on('exit')` event doesn't catch all the ways a process can exit.
+	The `process.on('exit')` event doesn't catch all the ways a process can exit.
 
-This package is useful for cleaning up before exiting.
+	This package is useful for cleaning up before exiting. To exit safely, call
+	`exitHook.exit()` instead of `process.exit()`.
 
-@param onExit - The callback function to execute when the process exits.
-@returns A function that removes the hook when called.
+	@param onExit - The callback function to execute when the process exits. If asynchronous, maxWait is required.
+	@param maxWait - An optional duration in ms to wait for onExit to complete. Required for asynchronous exit handlers.
+	@returns A function that removes the hook when called.
 
-@example
-```
-import exitHook from 'exit-hook';
+	@example
+	```
+	import exitHook from 'exit-hook';
 
-exitHook(() => {
-	console.log('Exiting');
-});
+	exitHook(() => {
+		console.log('Exiting');
+	});
 
-// You can add multiple hooks, even across files
-exitHook(() => {
-	console.log('Exiting 2');
-});
+	// You can add multiple hooks, even across files
+	// asynchronous hooks should include an amount of time to wait for
+	// their completion
+	exitHook(async () => {
+		console.log('Exiting 2');
+	}, 100);
 
-throw new Error('🦄');
+	throw new Error('🦄');
 
-//=> 'Exiting'
-//=> 'Exiting 2'
+	//=> 'Exiting'
+	//=> 'Exiting 2'
 
-// Removing an exit hook:
-const unsubscribe = exitHook(() => {});
+	// Removing an exit hook:
+	const unsubscribe = exitHook(() => {});
 
-unsubscribe();
-```
-*/
-export default function exitHook(onExit: () => void): () => void;
+	unsubscribe();
+	```
+	*/
+	(onExit: () => void | (() => Promise<void>), maxWait?: number): () => void;
```

Comment:
If I understand the intention, the `onExit` portion of this might need to be `onExit: () => void | Promise<void>`. I tried to provide an async function to `exitHook` and was told by TypeScript that the usage was wrong. I could be mistaken, but assumed this was allowed with these changes.

Reply:
This looks like its from the original commit e5f8afad. Looking at [HEAD d476a44a](https://github.com/sindresorhus/exit-hook/pull/20/files#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8R64), it's using `onExit: () => void | Promise<void>`.

Can you try against the latest commit and let me know if you're still seeing a weird TS definition? Appreciate the extra set of eyes!

Reply:
Latest version looks good. I'm not sure how I ended up with an outdated SHA.